File: remove-nth-node-from-end-of-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
  pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
    let mut sz = 0;
    {
      let mut cur = &head;
      while let Some(node) = cur {
        sz += 1;
        cur = &node.next;
      }
    }

    if n == sz {
      return head.unwrap().next;
    }

    let mut dummy = Some(Box::new(ListNode {
      val: 0,
      next: head,
    }));;
    
    let mut cur = &mut dummy;
    for _ in 0..(sz - n) {
      cur = &mut cur.as_mut().unwrap().next;
    }

    cur.as_mut().unwrap().next = cur.as_mut().unwrap().next.as_mut().unwrap().next.take();
    dummy.unwrap().next
  }
}

File: r/remove-nth-node-from-end-of-list.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    auto cur = head; int sz = 0;
    while (cur) cur = cur->next, ++sz;

    n = sz - n;
    if (n == 0) return head->next;

    cur = head;
    for (int it = 0; it < n - 1; ++it) cur = cur->next;
    cur->next = cur->next->next;
    return head;
  }
};
